File: Task2/Sources/Shapes.cpp
#include "Shapes.h"

// Очистка контейнера от элементов (освобождение памяти).
void Shapes::Clear() {
    for (int i = 0; i < lenght; ++i) {
        delete items[i];
    }
    lenght = 0;
}

// Ввод содержимого контейнера из указанного потока.
void Shapes::Input(ifstream& ifst) {
    while (!ifst.eof()) {
        //if ((items[lenght] = Shape::StaticIn(ifst)) != 0) {
        //    lenght++;
        //}

        int kind;
        ifst >> kind;

        double dencity;
        ifst >> dencity;

        switch (kind) {
            case 1:
                // Сфера.
            {
                // Прочитаем параметры фигуры.
                int radius;
                ifst >> radius;

                Shape *sphere = new Sphere(dencity, radius);
                items[lenght] = sphere;
            }
                break;
            case 2:
                // Параллелепипед.
            {
                // Прочитаем параметры фигуры.
                int a, b, c;
                ifst >> a >> b >> c;

                Shape *parallelepiped = new Parallelepiped(dencity, a, b, c);
                items[lenght] = parallelepiped;
            }
                break;
            case 3:
                // Правильный тетраэдр.
            {
                // Прочитаем параметры фигуры.
                int a;
                ifst >> a;

                Shape *rightTetrahedron = new RightTetrahedron(dencity, a);
                items[lenght] = rightTetrahedron;
            }
                break;
            default:
                // std::cout << "Some error during generation";
                break;
        }

        ++lenght;
    }
}

// Случайный ввод содержимого контейнера.
void Shapes::InputRandom(int size) {
    while (lenght < size) {
        // Сгенерируем тип фигуры.
        auto kind = rand() % 3 + 1;

        switch (kind) {
            case 1:
                // Сфера.
                {
                    // Сгенерируем параметры фигруы.
                    Shape *sphere = new Sphere();
                    items[lenght] = sphere;
                }
                break;
            case 2:
                // Параллелепипед.
                {
                    Shape *parallelepiped = new Parallelepiped();
                    items[lenght] = parallelepiped;
                }
                break;
            case 3:
                // Правильный тетраэдр.
                {
                    Shape *rightTetrahedron = new RightTetrahedron();
                    items[lenght] = rightTetrahedron;
                }
                break;
            default:
                // std::cout << "Some error during generation";
                break;
        }

        ++lenght;
    }
}

// Вывод содержимого контейнера в указанный поток.
void Shapes::Output(ofstream& ofst) {
    ofst << "Container contains " << lenght << " elements.\n";

    for (int i = 0; i < lenght; i++) {
        ofst << i + 1 << ") ";
        items[i]->Output(ofst);
    }
}

// Сортировка контейнера.
void Shapes::Quicksort(int first, int last) {
    int f = first;
    int l = last;
    Shape* mid = items[(f + l) / 2]; //вычисление опорного элемента

    do
    {
        while (items[f]->GetArea() > mid->GetArea()) {
            f++;
        }
        while (items[l]->GetArea() < mid->GetArea()) {
            l--;
        }

        if (f <= l) //перестановка элементов
        {
            Shape* temp = items[f];
            items[f] = items[l];
            items[l] = temp;
            f++;
            l--;
        }

    } while (f < l);

    if (first < l) {
        Quicksort(first, l);
    }
    if (f < last) {
        Quicksort(f, last);
    }
}

File: Task3/Sources/Shapes.py
import random
from Sphere import Sphere
from Parallelepiped import Parallelepiped
from RightTetrahedron import RightTetrahedron


class Shapes:
    length = None
    items = []

    # Конструктор.
    def __init__(self):
        self.length = 0

    # Деструктор.
    def __del__(self):
        self.clear()

    def clear(self):
        # del self.items
        print()

    # Ввод содержимого контейнера из указанного потока.
    def input(self, file):
        line = file.readline()

        while line:
            kind = line[0]
            data = file.readline().split(' ')
            density = data[0]

            if kind == '1':
                # Сфера.
                radius = data[1].replace("\n", "")
                sphere = Sphere(density, radius)
                self.items.append(sphere)
            else:
                if kind == '2':
                    # Параллелепипед.
                    a = data[1]
                    b = data[2]
                    c = data[3]
                    self.items.append(Parallelepiped(density, a, b, c))
                else:
                    if kind == '3':
                        # Правильный тетраэдр.
                        a = data[1]
                        self.items.append(RightTetrahedron(density, a))

            line = file.readline()
            self.length = self.length + 1

    # Случайный ввод содержимого контейнера.
    def input_random(self, size):
        while self.length < size:
            # Сгенерируем тип фигуры.
            kind = random.randint(1, 3)

            if kind == 1:
                # Сфера.
                self.items.append(Sphere())
            else:
                if kind == 2:
                    # Параллелепипед.
                    self.items.append(Parallelepiped())
                else:
                    if kind == 3:
                        # Правильный тетраэдр.
                        self.items.append(RightTetrahedron())
            self.length = self.length + 1

    def output(self, file):
        print("Container contains ", len(self.items), "elements.", file=file)

        counter = 1
        for i in self.items:
            print(counter, ")", i.output(), file=file)
            counter = counter + 1

    # Сортировка контейнера.
    def quicksort(self, first, last):
        f = first
        l = last
        mid = self.items[(f + l) // 2]  # вычисление опорного элемента

        condition = True
        while condition:
            while self.items[f].get_area() > mid.get_area():
                f = f + 1
            while self.items[l].get_area() < mid.get_area():
                l = l - 1

            if f <= l:
                temp = self.items[f]
                self.items[f] = self.items[l]
                self.items[l] = temp
                f = f + 1
                l = l - 1

            condition = f < l
        # Эмуляция do while

        if first < l:
            self.quicksort(first, l)
        if f < last:
            self.quicksort(f, last)
